### ISSUE TITLE
Provide a batching/chunking timer

### ISSUE THREAD

**Issue by oberstet** [crossbario/txaio#66]
Migrated from https://github.com/crossbario/crossbar/issues/722

**Comment by glyph** [comment]
crossbario/crossbar#722 mentions adding a new interface.  But it seems to me that this should just be an `IReactorTime.callLater` /  `IEventLoopJustKiddingAsyncioDoesntBelieveInFormalInterfaces.call_later` implementation, with no new interface defined…

**Comment by meejah** [comment]
So, the Autobahn libraries don't use asyncio or Twisted directly -- they go through a "lowest common denominator" sort of library called https://github.com/crossbario/txaio

That means, we want/need to provide an object that wraps over ``txaio.call_later`` and oberstet's idea from above means, basically, this: https://github.com/meejah/txaio/blob/issue66-batched-call-later/txaio/tx.py#L493

So, someone wishing to make "batched" instead of normal ``call_later`` calls would call ``txaio.make_batched_timer`` which will return a thing implementing ``IBatchedTimer`` (having just a ``.call_later`` method). So then any calls you want quantized with particular parameters would call ``.call_later`` on that object, with the same API as the underlying ``txaio.call_later`` (which is itself just a thin wrapper over ``IReactorTime.callLater`` or ``BaseEventLoop.call_later``.

**Comment by meejah** [comment]
...so if this were "plain" Twisted, I think we'd just have ``BatchedTimer`` which would take an ``IReactorTime`` as an argument (along with batching parameters) and ``BatchedTimer`` itself would just implement ``IReactorTime`` as well.

@glyph, is that what you were expecting?

**Comment by glyph** [comment]
> @glyph, is that what you were expecting?

More or less.  One minor nuance, though, based on advice from [The Talk](https://www.youtube.com/watch?v=3MNVP9-hglc); rather than actually exposing a `BatchedTimer` _class_, just expose a `batchedTimer` _function_, which promises only to return an `IReactorTime` provider.  Internally, of course, there probably is a `_BatchedTimer` class, but by refusing to export it as a public name, users of the library can't subclass it or try to override its internals.

**Comment by meejah** [comment]
So, in the current API I'm proposing, it sounds pretty similar to that: there's ``make_batched_timer`` that returns instances of an internal-only helper that happens to conform to an interface. See: https://github.com/crossbario/txaio/pull/67/files#diff-a5e5b34d1134c66bcfe8aa3ec80dac8fR27

Reasons for introducing ``IBatchedTimer`` as an interface are:

 - asyncio has no interface (e.g. like ``IReactorTime``)
 - we can't use Twisted interfaces directly, as txaio can be used *without* installing Twisted at all
 - make it more-clear we're not providing "full" Twisted API (e.g. ``IDelayedCall`` provides more than just ``.cancel`` on the returned objects, but the "lowest common denominator" here is that the "thing that ``call_later()`` returns" can only have ``.cancel()``)
 - txaio (for better or for worse) provides a module-level ``call_later`` API (instead of, e.g., trying to invent a "common" reactor/event-loop interface class) and has a "global" configured event-loop/reactor. It might be "better" to abstract the reactor/event-loop such that you can use txaio over multiple loops etc. ... but, it doesn't. Pragmatically: Twisted doesn't support multiple reactors (yet) anyway, and Autobahn doesn't need this.

So to summarize, before this change txaio had one API: ``txaio.call_later``. This change introduces a single method (``txaio.make_batched_timer``) which has just a ``.call_later`` method, with the same API as ``txaio.call_later`` (returns "a thing which has a ``.cancel``, and schedules a call for later).

**Comment by oberstet** [comment]
@meejah I also have the impression that what we now have is pretty much what @glyph is hinting at.

@glyph rgd The Talk:

I love the way those guys are talking .. in particular the "left one" .. like spilling out words in razor sharp grammar. Fun listening because it's so precise. I think I can do that in German, but sadly not in English=(

Then, the first "question" from the audience: "do you know zope.interface"? ;) Who's asking that?

Anyway: I grok that, composition over inheritance.

FWIW: we have that public API approach both in AutobahnJS and AutobahnCpp, not quite in AutobahnPython sadly. And not in AutobahnAndroid (but that is Java .. shudder .. at least it uses interfaces). This is my fault .. I got dragged away by Twisted's way of doing Protocols (inheritance) ;) Lame excuse, I know. How do you look at that from as of today?

In particular, we failed to at least do it differently for the WAMP layer in AutobahnPython ([this](https://github.com/crossbario/autobahn-python/blob/master/autobahn/wamp/interfaces.py#L258) and [this](https://github.com/crossbario/autobahn-python/blob/master/autobahn/wamp/protocol.py#L272) - users are supposed to subclass the latter to implement their app)

We do want to change that before going 1.0 in Autobahn though!

**Comment by oberstet** [comment]
* fixed via https://github.com/crossbario/txaio/pull/67
* released in 2.4.0
